File: kalaha/ai_engine.py
import random
from collections import OrderedDict
from typing import List, Tuple, Optional, Any, Dict

# Imports with fallback
try:
    from game_logic import (
        legal_moves, apply_move, is_terminal, evaluate,
        P1_PITS, P2_PITS, P1_STORE, P2_STORE, cleanup_board
    )
    from zobrist_hashing import zobrist
    from endgame_db import endgame_db
except ImportError:
    from kalaha.game_logic import (
        legal_moves, apply_move, is_terminal, evaluate,
        P1_PITS, P2_PITS, P1_STORE, P2_STORE, cleanup_board
    )
    from kalaha.zobrist_hashing import zobrist
    from kalaha.endgame_db import endgame_db

# Constants
MAX_DEPTH = 6
INF = float('inf')

# Transposition Table
TT: Dict[int, Tuple[float, int, str]] = {}

# Global counter for nodes visited
NODES_VISITED = 0

def evaluate_heuristic(board: List[int], player: int, strategy: str = 'balanced') -> float:
    """
    Advanced heuristic evaluation with multiple strategies.
    Positive value favors Player 0 (P1/Max).
    Strategies: 'balanced', 'aggressive', 'defensive', 'basic'
    """
    store_diff = board[P1_STORE] - board[P2_STORE]
    
    if strategy == 'basic':
        return float(store_diff)
        
    p1_side_seeds = sum(board[i] for i in P1_PITS)
    p2_side_seeds = sum(board[i] for i in P2_PITS)
    side_diff = p1_side_seeds - p2_side_seeds
    
    score = float(store_diff)
    
    if strategy == 'balanced':
        score += 0.5 * side_diff
    elif strategy == 'defensive':
        score += 0.8 * side_diff
        empty_p1 = sum(1 for i in P1_PITS if board[i] == 0)
        score -= 2.0 * empty_p1
    elif strategy == 'aggressive':
        score += 0.3 * side_diff
        
    return score

def order_moves(board: List[int], moves: List[int], player: int) -> List[int]:
    """
    Orders moves to improve Alpha-Beta pruning.
    """
    ordered = []
    
    for move in moves:
        sim_board, extra_turn = apply_move(board, move, player)
        
        score_diff = (sim_board[P1_STORE] - sim_board[P2_STORE]) if player == 0 else (sim_board[P2_STORE] - sim_board[P1_STORE])
        prev_diff = (board[P1_STORE] - board[P2_STORE]) if player == 0 else (board[P2_STORE] - board[P1_STORE])
        
        captured = (score_diff - prev_diff) > 1
        
        priority = 0.0
        if extra_turn:
            priority += 1000.0
        if captured:
            priority += 500.0
        
        priority += random.random()
        
        ordered.append((priority, move))
        
    ordered.sort(key=lambda x: x[0], reverse=True)
    return [m for p, m in ordered]

def alphabeta_tt_db(board: List[int], depth: int, alpha: float, beta: float, maximizing_player: bool, strategy: str = 'balanced') -> float:
    """
    Minimax with Alpha-Beta pruning, Transposition Table, and Endgame DB.
    """
    global NODES_VISITED
    NODES_VISITED += 1
    
    current_player = 0 if maximizing_player else 1
    
    # 1. Endgame DB Lookup (if seeds low enough)
    total_seeds = sum(board)
    if total_seeds <= max(10, endgame_db.max_seeds):
        exact_val = endgame_db.lookup(board, current_player)
        if exact_val is not None:
            return exact_val

    board_hash = zobrist.compute_hash(board, current_player)
    
    # 2. TT Lookup
    if board_hash in TT:
        tt_val, tt_depth, tt_flag = TT[board_hash]
        if tt_depth >= depth:
            if tt_flag == 'EXACT':
                return tt_val
            elif tt_flag == 'LOWERBOUND':
                alpha = max(alpha, tt_val)
            elif tt_flag == 'UPPERBOUND':
                beta = min(beta, tt_val)
            
            if alpha >= beta:
                return tt_val

    # 3. Terminal Node or Depth Limit
    if depth == 0 or is_terminal(board):
        if is_terminal(board):
            final_board = cleanup_board(board)
            val = float(final_board[P1_STORE] - final_board[P2_STORE])
            
            # Save solved terminal state to DB + TT
            endgame_db.add(board, current_player, int(val))
            TT[board_hash] = (val, 100, 'EXACT') 
            return val
        
        val = evaluate_heuristic(board, 0, strategy)
        TT[board_hash] = (val, depth, 'EXACT')
        return val
    
    possible_moves = legal_moves(board, current_player)
    ordered_moves = order_moves(board, possible_moves, current_player)
    
    value: float
    tt_flag: str = 'EXACT'

    if maximizing_player:
        value = -INF
        for move in ordered_moves:
            new_board, extra = apply_move(board, move, 0)
            
            if extra:
                score = alphabeta_tt_db(new_board, depth, alpha, beta, True, strategy)
            else:
                score = alphabeta_tt_db(new_board, depth - 1, alpha, beta, False, strategy)

            if score > value:
                value = score
            
            alpha = max(alpha, value)
            if alpha >= beta:
                tt_flag = 'LOWERBOUND'
                break
    else:
        value = INF
        for move in ordered_moves:
            new_board, extra = apply_move(board, move, 1)
            
            if extra:
                score = alphabeta_tt_db(new_board, depth, alpha, beta, False, strategy)
            else:
                score = alphabeta_tt_db(new_board, depth - 1, alpha, beta, True, strategy)
            
            if score < value:
                value = score
                
            beta = min(beta, value)
            if beta <= alpha:
                tt_flag = 'UPPERBOUND' 
                break

    TT[board_hash] = (value, depth, tt_flag)
    
    return value

def get_best_move(board: List[int], player: int, depth: int = MAX_DEPTH, strategy: str = 'balanced') -> Tuple[Optional[int], int]:
    """
    Determine the best move for the AI.
    Returns: (best_move, nodes_analyzed)
    """
    global NODES_VISITED
    NODES_VISITED = 0
    
    possible_moves = legal_moves(board, player)
    ordered_moves = order_moves(board, possible_moves, player)
    
    if not ordered_moves:
        return None, 0
        
    best_move = -1
    best_value = -INF if player == 0 else INF
        
    alpha = -INF
    beta = INF
    
    for move in ordered_moves:
        new_board, extra_turn = apply_move(board, move, player)
        
        if extra_turn:
            score = alphabeta_tt_db(new_board, depth, alpha, beta, player == 0, strategy)
        else:
            score = alphabeta_tt_db(new_board, depth - 1, alpha, beta, player != 0, strategy)
        
        if player == 0:
            if score > best_value:
                best_value = score
                best_move = move
            alpha = max(alpha, best_value)
        else:
            if score < best_value:
                best_value = score
                best_move = move
            beta = min(beta, best_value)
            
    return best_move, NODES_VISITED

File: kalaha-web/js/ai_engine.js
/**
 * AI Engine - Minimax with Alpha-Beta Pruning
 * Ported from Python (kalaha/ai_engine.py)
 */

import {
    legalMoves,
    applyMove,
    isTerminal,
    evaluate,
    P1_STORE,
    P2_STORE,
    P1_PITS,
    P2_PITS
} from './game_logic.js';

// Constants
const MAX_DEPTH = 6;
const INF = Infinity;

// Transposition Table
const TT = new Map();

// Global counter
let NODES_VISITED = 0;

/**
 * Heuristic evaluation with multiple strategies
 */
export function evaluateHeuristic(board, player, strategy = 'balanced') {
    const storeDiff = board[P1_STORE] - board[P2_STORE];

    if (strategy === 'basic') {
        return storeDiff;
    }

    const p1SideSeeds = P1_PITS.reduce((sum, i) => sum + board[i], 0);
    const p2SideSeeds = P2_PITS.reduce((sum, i) => sum + board[i], 0);
    const sideDiff = p1SideSeeds - p2SideSeeds;

    let score = storeDiff;

    if (strategy === 'balanced') {
        score += 0.5 * sideDiff;
    } else if (strategy === 'defensive') {
        score += 0.8 * sideDiff;
        const emptyP1 = P1_PITS.filter(i => board[i] === 0).length;
        score -= 2.0 * emptyP1;
    } else if (strategy === 'aggressive') {
        score += 0.3 * sideDiff;
    }

    return score;
}

/**
 * Order moves for better Alpha-Beta pruning
 */
function orderMoves(board, moves, player) {
    const scored = moves.map(move => {
        const result = applyMove(board, move, player);
        const simBoard = result.board;
        const extraTurn = result.extraTurn;

        const scoreDiff = player === 0
            ? (simBoard[P1_STORE] - simBoard[P2_STORE])
            : (simBoard[P2_STORE] - simBoard[P1_STORE]);
        const prevDiff = player === 0
            ? (board[P1_STORE] - board[P2_STORE])
            : (board[P2_STORE] - board[P1_STORE]);

        const captured = (scoreDiff - prevDiff) > 1;

        let priority = 0;
        if (extraTurn) priority += 1000;
        if (captured) priority += 500;
        priority += Math.random();

        return { move, priority };
    });

    scored.sort((a, b) => b.priority - a.priority);
    return scored.map(s => s.move);
}

/**
 * Simple hash function for board state
 */
function hashBoard(board) {
    return board.join(',');
}

/**
 * Minimax with Alpha-Beta Pruning
 */
function alphabeta(board, depth, alpha, beta, maximizing, strategy = 'balanced') {
    NODES_VISITED++;

    // Check transposition table
    const hash = hashBoard(board);
    if (TT.has(hash)) {
        const cached = TT.get(hash);
        if (cached.depth >= depth) {
            if (cached.flag === 'EXACT') return cached.value;
            if (cached.flag === 'LOWER' && cached.value > alpha) alpha = cached.value;
            if (cached.flag === 'UPPER' && cached.value < beta) beta = cached.value;
            if (alpha >= beta) return cached.value;
        }
    }

    // Terminal or depth limit
    if (isTerminal(board) || depth === 0) {
        const value = evaluateHeuristic(board, 0, strategy);
        TT.set(hash, { value, depth, flag: 'EXACT' });
        return value;
    }

    const player = maximizing ? 0 : 1;
    const moves = legalMoves(board, player);

    if (moves.length === 0) {
        const value = evaluateHeuristic(board, 0, strategy);
        TT.set(hash, { value, depth, flag: 'EXACT' });
        return value;
    }

    const orderedMoves = orderMoves(board, moves, player);

    if (maximizing) {
        let maxEval = -INF;

        for (const move of orderedMoves) {
            const result = applyMove(board, move, player);
            const newBoard = result.board;
            const extraTurn = result.extraTurn;

            const evalScore = alphabeta(newBoard, depth - 1, alpha, beta, !extraTurn, strategy);
            maxEval = Math.max(maxEval, evalScore);
            alpha = Math.max(alpha, evalScore);

            if (beta <= alpha) break; // Beta cutoff
        }

        const flag = maxEval <= alpha ? 'UPPER' : (maxEval >= beta ? 'LOWER' : 'EXACT');
        TT.set(hash, { value: maxEval, depth, flag });

        return maxEval;
    } else {
        let minEval = INF;

        for (const move of orderedMoves) {
            const result = applyMove(board, move, player);
            const newBoard = result.board;
            const extraTurn = result.extraTurn;

            const evalScore = alphabeta(newBoard, depth - 1, alpha, beta, extraTurn, strategy);
            minEval = Math.min(minEval, evalScore);
            beta = Math.min(beta, evalScore);

            if (beta <= alpha) break; // Alpha cutoff
        }

        const flag = minEval <= alpha ? 'UPPER' : (minEval >= beta ? 'LOWER' : 'EXACT');
        TT.set(hash, { value: minEval, depth, flag });

        return minEval;
    }
}

/**
 * Get best move for AI
 */
export function getBestMove(board, player, depth = MAX_DEPTH, strategy = 'balanced') {
    NODES_VISITED = 0;
    TT.clear();

    const moves = legalMoves(board, player);

    if (moves.length === 0) {
        return { move: null, nodes: NODES_VISITED };
    }

    if (moves.length === 1) {
        return { move: moves[0], nodes: NODES_VISITED };
    }

    const orderedMoves = orderMoves(board, moves, player);
    let bestMove = orderedMoves[0];
    let bestValue = player === 0 ? -INF : INF;

    for (const move of orderedMoves) {
        const result = applyMove(board, move, player);
        const newBoard = result.board;
        const extraTurn = result.extraTurn;

        const maximizing = player === 0 ? !extraTurn : extraTurn;
        const value = alphabeta(newBoard, depth - 1, -INF, INF, maximizing, strategy);

        if (player === 0) {
            if (value > bestValue) {
                bestValue = value;
                bestMove = move;
            }
        } else {
            if (value < bestValue) {
                bestValue = value;
                bestMove = move;
            }
        }
    }

    return { move: bestMove, nodes: NODES_VISITED };
}

/**
 * Get difficulty settings
 */
export function getDifficultySettings(difficulty) {
    const settings = {
        'easy': { depth: 2, strategy: 'basic' },
        'medium': { depth: 6, strategy: 'balanced' },
        'hard': { depth: 10, strategy: 'aggressive' }
    };

    return settings[difficulty] || settings['medium'];
}
